Support 32bit AS numbers

ShutterQuick opened the pull request:
The 16bit ASN limitation can be problematic when setting up Antrea in environments already heavily reliant on 32bit AS numbers.

EDIT: List of changes no longer correct. See the rest of the conversation.

This MR makes the following changes:
* Change ASN fields from int32 to uint32.
  As negative numbers are nonsensical this should be fine as far as I can figure, without breaking anything.
* In #6914 it is mentioned 32bit ASNs will have to wait until a new apiVersion.
  I don't know how the roadmap for this looks, but in case it's a while away I added a way to re-map 16bit numbers to 32bit numbers using annotations.
  While obviously a little nasty, it works just fine and could potentially be included as an undocumented or explicitly unsupported feature until a new schema version arrives in the future.

Example of use:

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: BGPPolicy
metadata:
  name: test-policy
  annotations:
    unsupported.bgp.antrea.io/local-asn-override: '4200100011'
    unsupported.bgp.antrea.io/peer-asn-overrides: '{ "10.0.0.1-65535": 4200100001 }'
spec:
  nodeSelector: {}
  localASN: 65535
  listenPort: 179
  advertisements:
    service:
      ipTypes: [ LoadBalancerIP ]
  bgpPeers:
    - address: 10.0.0.1
      asn: 65535
      port: 179
```

```
> antctl  get bgppolicy
NAME            ROUTER-ID     LOCAL-ASN  LISTEN-PORT
test-policy 10.0.0.1 4200100011 179

> antctl  get bgppeers
PEER              ASN        STATE
10.0.0.1:179 4200100001 Established
```

tnqn reviewed (commented):
@ShutterQuick thanks for the PR. If supporting 32-bit AS numbers is an immediate requirement, it would be preferable to introduce a new API version to handle it properly. And introducing a new API version with an updated field type should be a straightforward change.

cc @antoninbas @hongliangl

antoninbas commented:
Yes, @tnqn is correct. We had this discussion in https://github.com/antrea-io/antrea/pull/6914#discussion_r1910989608 and decided that a new API version would be needed. It's cleaner to define a `v1alpha2` version of the API than to rely on annotations (for which the deprecation process would not be clear). I assume we will need a conversion webhook. We already have an example in the code base (for the IPPool CRD) that you can follow: see `pkg/controller/ipam/convert.go`. When converting down from `v1alpha2` to `v1alpha1`, we can return an error status if the ASN doesn't fit in an `int32`.

ShutterQuick committed:
Implement support for 32-bit ASNs

Signed-off-by: Daniel Vassdal <daniel@vassdal.org>

ShutterQuick committed:
Add v1alpha2 CRD for BGPPolicy with support for 32-bit ASNs

Signed-off-by: Daniel Vassdal <daniel@vassdal.org>

ShutterQuick commented:
> If supporting 32-bit AS numbers is an immediate requirement, it would be preferable to introduce a new API version to handle it properly. And introducing a new API version with an updated field type should be a straightforward change.

Definitely more sensible, yeah :)

> I assume we will need a conversion webhook. We already have an example in the code base (for the IPPool CRD) that you can follow: see `pkg/controller/ipam/convert.go`. When converting down from `v1alpha2` to `v1alpha1`, we can return an error status if the ASN doesn't fit in an `int32`.

Thank you for the pointers! I've done it this way and updated the merge request.

I ran into an issue with returning an error from the convert hook, though.
Looks like Kubernetes really, really doesn't like it when a conversion hook returns an error.
At least in my testing, it insisted on converting to all versions with serve enabled on create and update.
It entered a very aggressive retry-fail-again loop trying to get it done, to no avail.
So it looks to me like returning error from the conversion hook is only sensible for transient errors.

For now I instead made it convert 32-bit ASNs to a 32-bit number when going from v1a2 to v1a1.
At least this preserves the information and doesn't violate the datatype. The range restriction on the other hand ...
Definitely not ideal, but I'm really not sure what would be a good solution here.

Maybe a feature flag to disable BGPPolicy v1alpha1?
Could then have an admission hook to
* Refuse admitting 32-bit ASNs when v1alpha1 bgppolicies are still in use
* Patch serve to false on v1alpha1 when admitting the first policy with 32-bit ASNs.
Urgh. There has to be a more sensible way ...

If any of you have any good ideas for how it should work please let me know I guess.

@tnqn @antoninbas

tnqn commented:
I did some tests and found kube-apiserver does call the conversion webhook regardless of which version I specify in the request and which version is the storage version. If failing a conversion request causes issues to client, could we use convert the value that cannot be accomodiated by int32 to an invalid value like 0, @antoninbas @ShutterQuick? In this way at least it's guaranteed any requests that previously worked continue to work. If users expect to use int64, they should switch to v1alpha2.

ShutterQuick commented:
@tnqn Thank you for having a look at it!

What you describe sounds reasonable to me. I can't think of anything better at least.
I've updated the PR and rebased it on main. The converter now replaces ASNs it can't convert with 0.

tnqn commented on pkg/controller/bgp/convert.go:
```diff
@@ -0,0 +1,97 @@
+					delete(confederation, "identifier")
+				}
+
+				memberASNs, _ := confederation["memberASNs"].([]interface{})
+				if len(memberASNs) == 0 {
+					// Deleting as the field has omitempty
+					delete(confederation, "memberASNs")
+				}
```
why is it needed?

ShutterQuick commented on pkg/agent/controller/bgp/controller.go:
```diff
@@ -410,6 +410,11 @@ func (c *Controller) syncBGPPolicy(ctx context.Context) error {
 			c.bgpPolicyState = nil
 		}
 
+		if ok, err := validatePolicy(effectivePolicy); !ok {
```
Though error on my side, will remove it 😅

ShutterQuick commented on pkg/controller/bgp/convert.go:
```diff
@@ -0,0 +1,97 @@
+		case "crd.antrea.io/v1alpha1":
+			localAsn, _, _ := unstructured.NestedInt64(convertedObject.Object, "spec", "localASN")
+			if localAsn > math.MaxUint16 {
+				// If confederation id is higher than we can represent we set it to 0.
+				// Not ideal but needed as conversion success is obligatory.
```
Oopsies, yeah that one isn't correct for that location for sure.

ShutterQuick commented on pkg/controller/bgp/convert.go:
```diff
@@ -0,0 +1,97 @@
+				// While removing the invalid confederation config would be easiest,
+				// we instead convert it, preserving 32-bit ASNs as 0.
+				// This is better than returning valid-looking objects that's missing configuration.
+				confederationId, _ := confederation["identifier"].(int64)
+				if confederationId > math.MaxUint16 {
```
Good catch, I screwed up the CRD, meant to make identifier 32-bit for v1alpha2.

tnqn commented on pkg/controller/bgp/convert.go:
```diff
@@ -0,0 +1,102 @@
+	}
+}
+
+func ConvertBgpPolicy(object *unstructured.Unstructured, toVersion string) (*unstructured.Unstructured, metav1.Status) {
```
```suggestion
func ConvertBGPPolicy(object *unstructured.Unstructured, toVersion string) (*unstructured.Unstructured, metav1.Status) {
```

tnqn commented on pkg/controller/bgp/convert.go:
```diff
@@ -0,0 +1,102 @@
+
+			confederation, ok, _ := unstructured.NestedMap(convertedObject.Object, "spec", "confederation")
+			if ok {
+				confederationId, _ := confederation["identifier"].(int64)
```
```suggestion
				confederationID, _ := confederation["identifier"].(int64)
```

tnqn commented on pkg/controller/bgp/convert.go:
```diff
@@ -0,0 +1,102 @@
+
+				memberASNs, _ := confederation["memberASNs"].([]interface{})
+				for i := range memberASNs {
+					memberAsn := memberASNs[i].(int64)
```
```suggestion
					memberASN := memberASNs[i].(int64)
```

tnqn commented on pkg/controller/bgp/convert_test.go:
```diff
@@ -0,0 +1,274 @@
+)
+
+func TestBGPPolicyConverter(t *testing.T) {
+	createV1a1Policy := func(localAsn int32, peerAsn int32, confederationId int32, confederationMemberASNs []int32) *crdv1alpha1.BGPPolicy {
```
```suggestion
	createV1a1Policy := func(localASN int32, peerASN int32, confederationID int32, confederationMemberASNs []int32) *crdv1alpha1.BGPPolicy {
```

tnqn commented on pkg/controller/bgp/convert_test.go:
```diff
@@ -0,0 +1,274 @@
+		return res
+	}
+
+	createV1a2Policy := func(localAsn int64, peerAsn int64, confederationId int64, confederationMemberASNs []int64) *crdv1alpha2.BGPPolicy {
```
ditto

tnqn commented on pkg/controller/bgp/convert_test.go:
```diff
@@ -0,0 +1,274 @@
+		return res
+	}
+
+	v1a1BgpPolicy16BitAsn := createV1a1Policy(65535, 65535, -1, []int32{})
```
ditto, so do other Asn

tnqn commented on build/charts/antrea/crds/bgppolicy.yaml:
```diff
@@ -2,12 +2,15 @@ apiVersion: apiextensions.k8s.io/v1
   versions:
     - name: v1alpha1
       served: true
-      storage: true
```
need to configure:
```
  conversion:
    strategy: Webhook
    webhook:
      conversionReviewVersions: ["v1", "v1beta1"]
      clientConfig:
        service:
          name: "antrea"
          namespace: "kube-system"
          path: "/convert/bgppolicy"
```
Otherwise the webhook won't be working

tnqn reviewed (commented):
LGTM overall, except that it’s missing the conversion configuration and has a few inconsistent initialism writings.

@antoninbas does the current approach (converting ASN greater than 65535 to 0 for v1alpha1) work for you?

ShutterQuick committed:
Run codegen

Signed-off-by: Daniel Vassdal <daniel@vassdal.org>
